src/main/java/org/springframework/data/r2dbc/function/ParameterbindingPreparedOperation.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.r2dbc.function;
+
+import io.r2dbc.spi.Connection;
+import io.r2dbc.spi.Statement;
+
+import java.util.Map;
+
+import org.springframework.data.r2dbc.domain.SettableValue;
+
+/**
+ * @author Jens Schauder
+ */
+public class ParameterbindingPreparedOperation implements PreparedOperation<BindableOperation> {
+
+	private final BindableOperation operation;
+	private final Map<String, SettableValue> byName;
+	private final Map<Integer, SettableValue> byIndex;
+
+	private ParameterbindingPreparedOperation(BindableOperation operation, Map<String, SettableValue> byName,
+			Map<Integer, SettableValue> byIndex) {
+
+		this.operation = operation;
+		this.byName = byName;
+		this.byIndex = byIndex;
+	}
+
+	ParameterbindingPreparedOperation(String sql, NamedParameterExpander namedParameters,
+			ReactiveDataAccessStrategy dataAccessStrategy, Map<String, SettableValue> byName,
+			Map<Integer, SettableValue> byIndex) {
+
+		this( //
+				namedParameters.expand(sql, dataAccessStrategy.getBindMarkersFactory(), new MapBindParameterSource(byName)), //
+				byName, //
+				byIndex //
+		);
+	}
+
+	@Override
+	public BindableOperation getSource() {
+		return operation;
+	}
+
+	@Override
+	public Statement bind(Connection connection) {
```

Comment:
`bind(…)` is somewhat misleading now. That would be rather `createBoundStatement()`. There is now no hook that would allow us to modify the SQL that is used to create the `Statement`. 

Right now, without #46, all this makes little sense.

Query Pre-processing would look like:

```java
QueryFilterFunction filter = ofPreparedOperation(preparedOperation -> {

  PreparedOperation operation = PreparedOperation.from(preparedOperation)
  .filterQuery(sql -> sql + " AND tenantId = :tenant")
  .withBinding("tenantId", "foobar")
  .build();

  return operation;
});

filter.and(ofPreparedOperation(operation -> {
  AUDIT_LOGGER.info("Executed Query: " + operation.toQuery());
  return operation;
}))
```